File: SOLID/Single.java
class Oficina {
    public void registrar() {
        System.out.println("Registrando entradas y salidas.");
    }
    public void guardarDatos() {
        System.out.println("Guardando datos del empleado.");
    }
    public void generarReporte() {
        System.out.println("Generando reporte.");
    }
}

class Checador {
    public void registrar() {
        System.out.println("Registrando entradas y salidas.");
    }
}

class Empleado {
    public void guardarDatos() {
        System.out.println("Guardando datos del empleado.");
    }
}

class GeneradorDeReportes {
    public void generarReporte() {
        System.out.println("Generando reporte.");
    }
}

public class Single {
    public static void main(String[] args) {
        Empleado empleado = new Empleado();
        Checador registro = new Checador();
        GeneradorDeReportes reporte = new GeneradorDeReportes();
        Oficina oficina = new Oficina();

        //mal
        oficina.guardarDatos();
        oficina.registrar();
        oficina.generarReporte();

        //bien
        empleado.guardarDatos();
        registro.registrar();
        reporte.generarReporte();
    }
}

File: SOLID/Single.cpp
#include <iostream>
using namespace std;

class Checador {
public:
    void registrar() {
        cout << "Registrando entradas y salidas." << endl;
    }
};

class Empleado {
public:
    void guardarDatos() {
        cout << "Guardando datos del empleado." << endl;
    }
};

class GeneradorDeReportes {
public:
    void generarReporte() {
        cout << "Generando reporte." << endl;
    }
};

//estaria mal
class Oficina{
public:
    void guardarDatos() {
        cout << "Guardando datos del empleado." << endl;
    }
    void registrar() {
        cout << "Registrando entradas y salidas." << endl;
    }
    void generarReporte() {
        cout << "Generando reporte." << endl;
    }
};

int main() {
    Empleado empleado;
    Checador registro;
    GeneradorDeReportes reporte;
    Oficina oficina;

    //estaria mal
    oficina.guardarDatos();
    oficina.registrar();
    oficina.generarReporte();

    //deberia ser
    empleado.guardarDatos();
    registro.registrar();
    reporte.generarReporte();

    return 0;
}
